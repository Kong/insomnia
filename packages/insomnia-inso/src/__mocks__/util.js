// eslint-disable-next-line filenames/match-exported
const mod = jest.requireActual('../util');

mod.exit = jest.fn();

module.exports = mod;
