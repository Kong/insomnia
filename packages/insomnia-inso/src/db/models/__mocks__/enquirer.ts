const __constructorMock = jest.fn();

let returnValue = 'test';

module.exports = {
  __constructorMock,
  __mockPromptRun: (value: string) => {
    returnValue = value;
  },
  AutoComplete: class {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any -- missing types from enquirer
    constructor(options: any) {
      __constructorMock(options);
    }

    run() {
      return returnValue;
    }
  },
};
