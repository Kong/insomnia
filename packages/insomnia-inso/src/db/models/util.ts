import { InsoError } from '../../errors';
import type { BaseModel } from './types';

export const matchIdIsh = ({ _id }: BaseModel, identifier: string) =>
  _id.startsWith(identifier);

export const generateIdIsh = ({ _id }: BaseModel, length = 10) =>
  _id.substr(0, length);

function indent(level: number, code: string, tab = '  |'): string {
  if (!level || level < 0) {
    return code;
  }

  const prefix = new Array(level + 1).join(tab);
  return `${prefix} ${code}`;
}

export const getDbChoice = (
  idIsh: string,
  message: string,
  config: {
        indent?: number
        hint?: string
    } = {},
) => ({
  name: idIsh,
  message: indent(config.indent || 0, message),
  value: `${message} - ${idIsh}`,
  hint: config.hint || `${idIsh}`,
});

export const ensureSingleOrNone = <T>(
  items: T[],
  entity: string,
): T | null | undefined => {
  if (items.length === 1) {
    return items[0];
  }

  if (items.length === 0) {
    return null;
  }

  throw new InsoError(
    `Expected single or no ${entity} in the data store, but found multiple (${items.length}).`,
  );
};
export const ensureSingle = <T>(items: T[], entity: string): T => {
  if (items.length === 1) {
    return items[0];
  }

  if (items.length === 0) {
    throw new InsoError(
      `Expected single ${entity} in the data store, but found none.`,
    );
  }

  throw new InsoError(
    `Expected single ${entity} in the data store, but found multiple (${items.length}).`,
  );
};
