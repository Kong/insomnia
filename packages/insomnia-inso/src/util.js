// @flow
import * as packageJson from '../package.json';
import { handleError } from './errors';

export function getVersion() {
  return isDevelopment() ? 'dev' : packageJson.version;
}

export function isDevelopment() {
  return process.env.NODE_ENV === 'development';
}

export function logErrorExit1(err?: Error) {
  handleError(err);
  process.exit(1);
}

export async function exit(result: Promise<boolean>): Promise<void> {
  return result.then(r => process.exit(r ? 0 : 1)).catch(logErrorExit1);
}

export function getDefaultAppDataDir(): string {
  const dir = process.env.DEFAULT_APP_DATA_DIR;

  if (!dir) {
    throw new Error('Environment variable DEFAULT_APP_DATA_DIR is not set.');
  }

  return dir;
}
