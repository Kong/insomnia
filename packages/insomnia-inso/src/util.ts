import packageJson from '../package.json';
import { handleError } from './errors';

export const getVersion = () => {
  return isDevelopment() ? 'dev' : packageJson.version;
};

export const isDevelopment = () => {
  return process.env.NODE_ENV === 'development';
};

export const logErrorExit1 = (err?: Error) => {
  handleError(err);
  process.exit(1);
};

export const exit = async (result: Promise<boolean>): Promise<void> => {
  return result.then(r => process.exit(r ? 0 : 1)).catch(logErrorExit1);
};

export const getDefaultAppName = (): string => {
  const name = process.env.DEFAULT_APP_NAME;

  if (!name) {
    throw new Error('Environment variable DEFAULT_APP_NAME is not set.');
  }

  return name;
};

/**
 * Extract an object from a string array with a key=value format
 * (e.g var arr = ["prop1=value", "prop2=value" ...])
 * @param data An array of 'equal'(=) separated key value pairs
 */
export const parseObjFromKeyValuePair = (data: string[] = []): Record<string, unknown> => {
  // Placeholder data
  const obj: Record<string, unknown> = {};
  // Cycling for extraction each pair key-value
  data.forEach(arg => {
    // Extract option key-value
    const key = arg.substring(0, arg.indexOf('='));
    let value: unknown = arg.substring(arg.indexOf('=') + 1);
    // Everything is treated as a string, except:
    if (value === 'true' || value === 'false') value = value === 'true';
    else if (value === 'null') value = null;
    else if (value === 'undefined') value = undefined;
    else if (!isNaN(<number>value)) value = Number(value);
    // Assign
    obj[key] = value;
  });
  return obj;
};

// eslint-disable-next-line @typescript-eslint/no-empty-function
export const noop = () => {};
