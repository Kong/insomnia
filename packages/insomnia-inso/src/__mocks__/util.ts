import { jest } from '@jest/globals';

import * as utilOriginal from '../util';

// eslint-disable-next-line filenames/match-exported
const util = jest.requireActual('../util') as typeof utilOriginal;
util.exit = jest.fn();

module.exports = util;
