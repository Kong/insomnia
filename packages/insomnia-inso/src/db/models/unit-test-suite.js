// @flow
import type { Database } from '../index';
import type { UnitTestSuite } from './types';
import { loadApiSpec } from './api-spec';
import { mustFindSingleOrNone } from '../index';
import { AutoComplete } from 'enquirer';
import flattenDeep from 'lodash.flattendeep';
import { generateIdIsh, getDbChoice, matchIdIsh } from './util';
import { loadWorkspace } from './workspace';

export const loadTestSuites = (db: Database, identifier: string): Array<UnitTestSuite> => {
  const apiSpec = loadApiSpec(db, identifier);
  const workspace = loadWorkspace(db, apiSpec?.parentId || identifier);

  // if identifier is for an apiSpec or a workspace, return all suites for that workspace
  if (workspace) {
    return db.UnitTestSuite.filter(s => s.parentId === workspace._id);
  }

  // Identifier is for one specific suite; find it
  const result = mustFindSingleOrNone(
    db.UnitTestSuite,
    s => matchIdIsh(s, identifier) || s.name === identifier,
  );

  return result ? [result] : [];
};

export const promptTestSuites = async (
  db: Database,
  ci: boolean,
): Promise<Array<UnitTestSuite>> => {
  if (ci) {
    return [];
  }

  const choices = db.ApiSpec.map(spec => [
    getDbChoice(generateIdIsh(spec), spec.fileName),
    ...db.UnitTestSuite.filter(suite => suite.parentId === spec.parentId).map(suite =>
      getDbChoice(generateIdIsh(suite), suite.name, { indent: 1 }),
    ),
  ]);

  if (!choices.length) {
    return [];
  }

  const prompt = new AutoComplete({
    name: 'testSuite',
    message: 'Select a document or unit test suite',
    choices: flattenDeep(choices),
  });

  const [idIsh] = (await prompt.run()).split(' - ').reverse();
  return loadTestSuites(db, idIsh);
};
