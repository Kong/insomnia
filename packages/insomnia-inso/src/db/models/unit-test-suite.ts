// @ts-expect-error the enquirer types are incomplete https://github.com/enquirer/enquirer/pull/307
import { AutoComplete } from 'enquirer';
import flattenDeep from 'lodash.flattendeep';

import { logger } from '../../logger';
import type { Database } from '../index';
import { loadApiSpec } from './api-spec';
import type { UnitTestSuite } from './types';
import {
  ensureSingleOrNone,
  generateIdIsh,
  getDbChoice,
  matchIdIsh,
} from './util';
import { loadWorkspace } from './workspace';

export const loadUnitTestSuite = (
  db: Database,
  identifier: string,
): UnitTestSuite | null | undefined => {
  // Identifier is for one specific suite; find it
  logger.trace(
    'Load unit test suite with identifier `%s` from data store',
    identifier,
  );
  const items = db.UnitTestSuite.filter(
    suite => matchIdIsh(suite, identifier) || suite.name === identifier,
  );
  logger.trace('Found %d.', items.length);
  return ensureSingleOrNone(items, 'unit test suite');
};
export const loadTestSuites = (
  db: Database,
  identifier: string,
): UnitTestSuite[] => {
  const apiSpec = loadApiSpec(db, identifier);
  const workspace = loadWorkspace(db, apiSpec?.parentId || identifier); // if identifier is for an apiSpec or a workspace, return all suites for that workspace

  if (workspace) {
    return db.UnitTestSuite.filter(s => s.parentId === workspace._id);
  } // load particular suite

  const result = loadUnitTestSuite(db, identifier);
  return result ? [result] : [];
};
export const promptTestSuites = async (
  db: Database,
  ci: boolean,
): Promise<UnitTestSuite[]> => {
  if (ci) {
    return [];
  }

  const choices = db.ApiSpec.map(spec => [
    getDbChoice(generateIdIsh(spec), spec.fileName),
    ...db.UnitTestSuite.filter(
      suite => suite.parentId === spec.parentId,
    ).map(suite =>
      getDbChoice(generateIdIsh(suite), suite.name, {
        indent: 1,
      }),
    ),
  ]);

  if (!choices.length) {
    return [];
  }

  const prompt = new AutoComplete({
    name: 'testSuite',
    message: 'Select a document or unit test suite',
    choices: flattenDeep(choices),
  });
  logger.trace('Prompt for document or test suite');
  const [idIsh] = (await prompt.run()).split(' - ').reverse();
  return loadTestSuites(db, idIsh);
};
