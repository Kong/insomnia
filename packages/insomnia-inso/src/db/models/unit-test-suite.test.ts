import enquirer from 'enquirer';

import type { Database } from '../index';
import { emptyDb } from '../index';
import type { ApiSpec, UnitTestSuite, Workspace } from './types';
import { loadTestSuites, promptTestSuites } from './unit-test-suite';
import { generateIdIsh } from './util';

jest.mock('enquirer');

describe('Unit Test Suite', () => {
  let db: Database = emptyDb();

  const workspace: Partial<Workspace> = {
    _id: 'wrk_1234567890',
    name: 'workspace name',
  };

  const spec: Partial<ApiSpec> = {
    _id: 'spc_1234567890',
    fileName: 'spec name',
    parentId: workspace._id,
  };

  const suite1: Partial<UnitTestSuite> = {
    _id: 'uts_1234567890',
    name: 'suite one',
    parentId: workspace._id,
  };

  const suite2: Partial<UnitTestSuite> = {
    _id: 'uts_987654321',
    name: 'suite two',
    parentId: workspace._id,
  };

  beforeEach(() => {
    db = emptyDb();
    const dummySuite: Partial<UnitTestSuite> = {
      _id: 'uts_dummy',
      name: 'dummy suite',
      parentId: 'dummy parent',
    };
    db.Workspace.push(workspace);
    db.ApiSpec.push(spec);
    db.UnitTestSuite.push(suite1);
    db.UnitTestSuite.push(suite2);
    db.UnitTestSuite.push(dummySuite);
    jest.clearAllMocks();
  });

  describe('promptTestSuites()', () => {
    it('should return empty array if ci', () => {
      expect(promptTestSuites(db, true)).resolves.toHaveLength(0);
    });

    it('should return empty array if no specs', () => {
      db.ApiSpec = [];
      expect(promptTestSuites(db, false)).resolves.toHaveLength(0);
    });

    it('should return empty array if prompt result does not contain closing [ - id]', async () => {
      enquirer.__mockPromptRun('malformed result');

      expect(promptTestSuites(db, false)).resolves.toHaveLength(0);
    });

    it('should load suite 1 after prompt result', async () => {
      enquirer.__mockPromptRun('suite one - uts_123456');

      const result = await promptTestSuites(db, false);
      expect(result).toHaveLength(1);
      expect(result).toContain(suite1);
    });

    it('should match snapshot of autocomplete config', async () => {
      await promptTestSuites(db, false);
      expect(enquirer.__constructorMock.mock.calls[0][0]).toMatchSnapshot();
    });
  });
  describe('loadTestSuites()', () => {
    it('should return empty array if no suites matched', () => {
      expect(loadTestSuites(db, 'not-found')).toHaveLength(0);
    });

    it.each([generateIdIsh(workspace), workspace._id, workspace.name])(
      'should load all suites that match by workspace id: %o',
      (identifier) => {
        const result = loadTestSuites(db, identifier);
        expect(result).toHaveLength(2);
        expect(result).toContain(suite1);
        expect(result).toContain(suite2);
      },
    );

    it.each([generateIdIsh(spec), spec._id, spec.fileName])(
      'should load all suites that match by apiSpec id: %o',
      (identifier) => {
        const result = loadTestSuites(db, identifier);
        expect(result).toHaveLength(2);
        expect(result).toContain(suite1);
        expect(result).toContain(suite2);
      },
    );

    it.each([generateIdIsh(suite1), suite1._id, suite1.name])(
      'should load single suite that matches by id: %o',
      (identifier) => {
        const result = loadTestSuites(db, identifier);
        expect(result).toHaveLength(1);
        expect(result).toContain(suite1);
      },
    );
  });
});
