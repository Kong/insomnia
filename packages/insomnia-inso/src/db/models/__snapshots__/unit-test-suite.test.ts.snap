// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit Test Suite promptTestSuites() should match snapshot of autocomplete config 1`] = `
Object {
  "choices": Array [
    Object {
      "hint": "spc_123456",
      "message": "spec name",
      "name": "spc_123456",
      "value": "spec name - spc_123456",
    },
    Object {
      "hint": "uts_123456",
      "message": "  | suite one",
      "name": "uts_123456",
      "value": "suite one - uts_123456",
    },
    Object {
      "hint": "uts_987654",
      "message": "  | suite two",
      "name": "uts_987654",
      "value": "suite two - uts_987654",
    },
  ],
  "message": "Select a document or unit test suite",
  "name": "testSuite",
}
`;
